File: social-validation-engine/modules/validation_engine.py
"""
modules/validation_engine.py — Module 8 (TDD §2.8)

Fully DETERMINISTIC scoring layer — no AI call.
Takes structured outputs of all previous modules and produces:
  - validation_score: 0-100
  - verdict: 'Strong opportunity' | 'Needs more digging' | 'Weak signal'
  - reasoning: plain-English explanation of the score

Weights from config.py / TDD §4:
  pain_point_frequency  30%
  buying_intent         25%
  competitor_weakness   20%
  feature_demand        15%
  market_activity       10%

Verdict bands (TDD §4):
  70-100 → Strong opportunity
  40-69  → Needs more digging
  0-39   → Weak signal
"""
from __future__ import annotations

import math
import logging
from typing import Any

from config import SCORING_WEIGHTS, VERDICT_BANDS

log = logging.getLogger(__name__)

# ── Sub-score normalisation helpers ──────────────────────────────────────────

def _sigmoid_normalize(value: float, midpoint: float = 50.0, steepness: float = 0.05) -> float:
    """Maps any positive float onto [0, 1] using a sigmoid curve."""
    return 1 / (1 + math.exp(-steepness * (value - midpoint)))


def _clamp(value: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, value))


# ── Factor scorers (each returns a float in [0, 1]) ──────────────────────────

def _score_pain_point_frequency(
    pain_points: list[dict[str, Any]],
) -> float:
    """
    Normalised sum of mention counts, weighted by severity.
    Baseline: 100 weighted mentions = 1.0 score.
    """
    if not pain_points:
        return 0.0
    weighted = sum(
        pp.get("mentions", 0) * (pp.get("severity") or 3) / 5
        for pp in pain_points
    )
    return _clamp(_sigmoid_normalize(weighted, midpoint=50, steepness=0.04))


def _score_buying_intent(
    sentiment: dict[str, Any],
) -> float:
    """
    buying_intent_count / total_tagged.  Bonus for active_search signals.
    """
    total = sentiment.get("total_tagged", 0)
    if total == 0:
        return 0.1   # slight baseline — no data ≠ definitely zero demand
    bi = sentiment.get("buying_intent_count", 0)
    active = sentiment.get("active_search_count", 0)
    ratio = (bi + 0.5 * active) / total
    return _clamp(ratio)


def _score_competitor_weakness(
    pain_points: list[dict[str, Any]],
    competitors: list[dict[str, Any]],
) -> float:
    """
    Measures the gap: what pain points are NOT covered by any competitor's features?
    More uncovered pain → more opportunity.
    """
    if not competitors:
        # No known competitors = either blue ocean or no validated market yet
        # Neutral score — let other factors decide
        return 0.5

    all_competitor_features = set()
    for c in competitors:
        for f in (c.get("missing_features") or []):
            all_competitor_features.add(f.lower())

    pain_labels = [pp["pain_point"].lower() for pp in pain_points]
    uncovered = sum(
        1 for label in pain_labels
        if not any(cf in label or label in cf for cf in all_competitor_features)
    )
    if not pain_labels:
        return 0.3
    gap_ratio = uncovered / len(pain_labels)
    return _clamp(gap_ratio)


def _score_feature_demand(
    features: list[dict[str, Any]],
) -> float:
    """
    Total feature mention count, normalised.
    Baseline: 30 total feature mentions = 1.0.
    """
    total = sum(f.get("mentions", 0) for f in features)
    if total == 0:
        return 0.1
    return _clamp(_sigmoid_normalize(total, midpoint=30, steepness=0.08))


def _score_market_activity(
    sources: list[dict[str, Any]],
) -> float:
    """
    Volume + recency of collected posts.
    Baseline: 200 posts = 1.0 volume score.
    Recency: fraction of posts in the last 180 days.
    """
    from datetime import datetime, timezone, timedelta
    if not sources:
        return 0.0
    volume_score = _clamp(_sigmoid_normalize(len(sources), midpoint=100, steepness=0.02))

    cutoff = datetime.now(tz=timezone.utc) - timedelta(days=180)
    recent = 0
    for s in sources:
        posted = s.get("posted_at")
        if posted:
            try:
                if isinstance(posted, str):
                    from datetime import datetime as dt
                    ts = dt.fromisoformat(posted.replace("Z", "+00:00"))
                else:
                    ts = posted
                if ts > cutoff:
                    recent += 1
            except ValueError:
                pass
    recency_score = recent / len(sources) if sources else 0.0
    return _clamp(0.6 * volume_score + 0.4 * recency_score)


# ── Verdict band lookup ───────────────────────────────────────────────────────

def _verdict_from_score(score: int) -> str:
    for lo, hi, label in VERDICT_BANDS:
        if lo <= score <= hi:
            return label
    return "Weak signal"


# ── Main entry point ──────────────────────────────────────────────────────────

def compute_validation_score(
    pain_points: list[dict[str, Any]],
    sentiment: dict[str, Any],
    competitors: list[dict[str, Any]],
    features: list[dict[str, Any]],
    sources: list[dict[str, Any]],
) -> dict[str, Any]:
    """
    Returns: {validation_score (int 0-100), verdict (str), reasoning (str)}
    """
    f1 = _score_pain_point_frequency(pain_points)
    f2 = _score_buying_intent(sentiment)
    f3 = _score_competitor_weakness(pain_points, competitors)
    f4 = _score_feature_demand(features)
    f5 = _score_market_activity(sources)

    w = SCORING_WEIGHTS
    raw_score = (
        f1 * w["pain_point_frequency"]
        + f2 * w["buying_intent"]
        + f3 * w["competitor_weakness"]
        + f4 * w["feature_demand"]
        + f5 * w["market_activity"]
    )
    # Scale 0-1 → 0-100
    validation_score = round(_clamp(raw_score) * 100)
    verdict = _verdict_from_score(validation_score)

    # Build plain-English reasoning
    pain_count = len(pain_points)
    bi_pct = round(sentiment.get("buying_intent_count", 0) / max(sentiment.get("total_tagged", 1), 1) * 100)
    competitor_count = len(competitors)
    feature_count = len(features)
    source_count = len(sources)

    reasoning = (
        f"Score: {validation_score}/100 — {verdict}.\n\n"
        f"Pain-point frequency ({round(w['pain_point_frequency']*100)}% weight, sub-score {round(f1*100)}/100): "
        f"{pain_count} distinct pain points found across {source_count} posts.\n"
        f"Buying intent ({round(w['buying_intent']*100)}% weight, sub-score {round(f2*100)}/100): "
        f"{bi_pct}% of tagged posts show explicit buying intent or active tool-searching.\n"
        f"Competitor weakness ({round(w['competitor_weakness']*100)}% weight, sub-score {round(f3*100)}/100): "
        f"{competitor_count} competitor(s) found; gap analysis shows {'significant' if f3 > 0.6 else 'moderate' if f3 > 0.3 else 'limited'} unmet demand.\n"
        f"Feature demand ({round(w['feature_demand']*100)}% weight, sub-score {round(f4*100)}/100): "
        f"{feature_count} distinct feature request(s) surfaced from competitor reviews.\n"
        f"Market activity ({round(w['market_activity']*100)}% weight, sub-score {round(f5*100)}/100): "
        f"{source_count} posts collected from Reddit."
    )

    log.info("validation_engine: score=%d verdict='%s'", validation_score, verdict)
    return {
        "validation_score": validation_score,
        "verdict": verdict,
        "reasoning": reasoning,
    }

File: apps/backend/src/tools/social-validation-engine/services/validation-engine.ts
import { SCORING_WEIGHTS, VERDICT_BANDS } from '../config';

function sigmoidNormalize(value: number, midpoint = 50.0, steepness = 0.05): number {
  return 1 / (1 + Math.exp(-steepness * (value - midpoint)));
}

function clamp(value: number, lo = 0.0, hi = 1.0): number {
  return Math.max(lo, Math.min(hi, value));
}

function scorePainPointFrequency(painPoints: any[]): number {
  if (!painPoints || painPoints.length === 0) return 0.0;
  const weighted = painPoints.reduce((sum, pp) => {
    return sum + (pp.mentions || 0) * (pp.severity || 3) / 5;
  }, 0);
  return clamp(sigmoidNormalize(weighted, 50, 0.04));
}

function scoreBuyingIntent(sentiment: any): number {
  const total = sentiment.total_tagged || 0;
  if (total === 0) return 0.1; // slight baseline
  const bi = sentiment.buying_intent_count || 0;
  const active = sentiment.active_search_count || 0;
  const ratio = (bi + 0.5 * active) / total;
  return clamp(ratio);
}

function scoreCompetitorWeakness(painPoints: any[], competitors: any[]): number {
  if (!competitors || competitors.length === 0) {
    return 0.5; // Neutral
  }

  const allCompetitorFeatures = new Set<string>();
  for (const c of competitors) {
    const missing = c.missing_features || [];
    for (const f of missing) {
      allCompetitorFeatures.add(f.toLowerCase());
    }
  }

  const painLabels = painPoints.map(pp => (pp.pain_point || '').toLowerCase());
  if (painLabels.length === 0) return 0.3;

  // Enforce gap analysis: count pain points that ARE mentioned in missing features
  // (meaning competitors are missing a solution for this pain point)
  let uncovered = 0;
  for (const label of painLabels) {
    const isUncovered = Array.from(allCompetitorFeatures).some(cf => {
      return cf.includes(label) || label.includes(cf);
    });
    if (isUncovered) {
      uncovered++;
    }
  }

  const gapRatio = uncovered / painLabels.length;
  return clamp(gapRatio);
}

function scoreFeatureDemand(features: any[]): number {
  if (!features || features.length === 0) return 0.1;
  const total = features.reduce((sum, f) => sum + (f.mentions || 0), 0);
  return clamp(sigmoidNormalize(total, 30, 0.08));
}

function scoreMarketActivity(sources: any[]): number {
  if (!sources || sources.length === 0) return 0.0;
  const volumeScore = clamp(sigmoidNormalize(sources.length, 100, 0.02));

  // Recency: fraction of posts in last 180 days
  const cutoff = Date.now() - 180 * 24 * 60 * 60 * 1000;
  let recent = 0;
  for (const s of sources) {
    const posted = s.posted_at;
    if (posted) {
      try {
        const ts = new Date(posted).getTime();
        if (ts > cutoff) recent++;
      } catch {}
    }
  }

  const recencyScore = recent / sources.length;
  return clamp(0.6 * volumeScore + 0.4 * recencyScore);
}

function getVerdictFromScore(score: number): string {
  for (const [lo, hi, label] of VERDICT_BANDS) {
    if (score >= lo && score <= hi) {
      return label;
    }
  }
  return "Weak signal";
}

export function computeValidationScore(
  painPoints: any[],
  sentiment: any,
  competitors: any[],
  features: any[],
  sources: any[]
): { validation_score: number; verdict: string; reasoning: string } {
  const f1 = scorePainPointFrequency(painPoints);
  const f2 = scoreBuyingIntent(sentiment);
  const f3 = scoreCompetitorWeakness(painPoints, competitors);
  const f4 = scoreFeatureDemand(features);
  const f5 = scoreMarketActivity(sources);

  const w = SCORING_WEIGHTS;
  const rawScore =
    f1 * w.pain_point_frequency +
    f2 * w.buying_intent +
    f3 * w.competitor_weakness +
    f4 * w.feature_demand +
    f5 * w.market_activity;

  const validationScore = Math.round(clamp(rawScore) * 100);
  const verdict = getVerdictFromScore(validationScore);

  const biPct = Math.round((sentiment.buying_intent_count || 0) / Math.max(sentiment.total_tagged || 1, 1) * 100);

  const reasoning = `Score: ${validationScore}/100 — ${verdict}.\n\n` +
    `Pain-point frequency (${Math.round(w.pain_point_frequency * 100)}% weight, sub-score ${Math.round(f1 * 100)}/100): ` +
    `${painPoints.length} distinct pain points found across ${sources.length} posts.\n` +
    `Buying intent (${Math.round(w.buying_intent * 100)}% weight, sub-score ${Math.round(f2 * 100)}/100): ` +
    `${biPct}% of tagged posts show explicit buying intent or active tool-searching.\n` +
    `Competitor weakness (${Math.round(w.competitor_weakness * 100)}% weight, sub-score ${Math.round(f3 * 100)}/100): ` +
    `${competitors.length} competitor(s) found; gap analysis shows ${f3 > 0.6 ? 'significant' : f3 > 0.3 ? 'moderate' : 'limited'} unmet demand.\n` +
    `Feature demand (${Math.round(w.feature_demand * 100)}% weight, sub-score ${Math.round(f4 * 100)}/100): ` +
    `${features.length} distinct feature request(s) surfaced from competitor reviews.\n` +
    `Market activity (${Math.round(w.market_activity * 100)}% weight, sub-score ${Math.round(f5 * 100)}/100): ` +
    `${sources.length} posts collected from Reddit.`;

  return {
    validation_score: validationScore,
    verdict,
    reasoning
  };
}
